rust/connlib/tunnel/src/sockets.rs

```diff
@@ -68,42 +68,59 @@ impl Sockets {
         self.socket_v6.as_ref().map(|s| s.socket.as_raw_fd())
     }
 
-    pub fn poll_send_ready(&mut self, cx: &mut Context<'_>) -> Poll<io::Result<()>> {
-        if let Some(socket) = self.socket_v4.as_mut() {
+    pub fn poll_send_ready(&self, cx: &mut Context<'_>) -> Poll<io::Result<()>> {
+        if let Some(socket) = self.socket_v4.as_ref() {
             ready!(socket.poll_send_ready(cx))?;
         }
 
-        if let Some(socket) = self.socket_v6.as_mut() {
+        if let Some(socket) = self.socket_v6.as_ref() {
             ready!(socket.poll_send_ready(cx))?;
         }
 
         Poll::Ready(Ok(()))
     }
 
-    pub fn try_send(&mut self, transmit: &Transmit) -> Result<usize> {
+    pub fn try_send(&self, transmit: &Transmit) -> io::Result<usize> {
         match transmit.dst {
             SocketAddr::V4(_) => {
-                let socket = self.socket_v4.as_ref().ok_or(Error::NoIpv4)?;
+                let socket = self.socket_v4.as_ref().ok_or(io::Error::new(
+                    io::ErrorKind::NotConnected,
+                    "no IPv4 socket",
+                ))?;
                 Ok(socket.try_send_to(transmit.src, transmit.dst, &transmit.payload)?)
             }
             SocketAddr::V6(_) => {
-                let socket = self.socket_v6.as_ref().ok_or(Error::NoIpv6)?;
+                let socket = self.socket_v6.as_ref().ok_or(io::Error::new(
+                    io::ErrorKind::NotConnected,
+                    "no IPv6 socket",
+                ))?;
                 Ok(socket.try_send_to(transmit.src, transmit.dst, &transmit.payload)?)
             }
         }
     }
 
-    pub fn poll_recv_from<'a>(
-        &'a mut self,
+    pub fn poll_recv_from<'b>(
+        &self,
+        buffer: &'b mut [u8],
         cx: &mut Context<'_>,
-    ) -> Poll<io::Result<impl Iterator<Item = Received<'a>>>> {
+    ) -> Poll<io::Result<impl Iterator<Item = Received<'b>>>> {
+        let (ip4_buf, ip6_buf) = buffer.split_at_mut(buffer.len() / 2);
```

Comment:
This api is a bit anti-intuitive, when passing the buffer I'd expect that the whole thing can be used to receive packets. I think we should just pass 2 buffers in the signature.

Reply:
Yeah I wasn't quite sure how to best deal with that. It also felt weird taking two buffers but perhaps that is better because it is less surprising.

I'll change it and split the buffer further up.